### PR TITLE
Move linting into ink repo

### PR DESCRIPTION
**Pull request by athei**
In preparation for the changes to `cargo contract` which will make the lints being build as part of the contract build we move it here. The lints are tightly coupled to the ink! code. There is no reason why they should exist in the `cargo contract` repo.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
## 🦑 📈 ink! Example Contracts ‒ Changes Report 📉 🦑
 These are the results when building the `examples/*` contracts from this branch with `cargo-contract 1.5.0-ef06f4d` and comparing them to ink! `master`: 

|                       | Δ Optimized Size | Δ Used Gas | Total Optimized Size | Total Used Gas |
|:-----------------------|------------------:|------------:|----------------------:|----------------:|
| `accumulator`         |                  |            | 1.00 K               |                |
| `adder`               |                  |            | 2.04 K               |                |
| `contract-terminate`  |                  |            | 0.92 K               | 275_000        |
| `contract-transfer`   |                  |            | 8.36 K               | 75_000         |
| `delegate-calls`      |                  |            | 2.89 K               | 76_242         |
| `delegator`           |                  |            | 6.34 K               | 232_138        |
| `dns`                 |                  |            | 8.81 K               | 225_000        |
| `erc1155`             |                  |            | 17.15 K              | 450_000        |
| `erc20`               |                  |            | 8.42 K               | 225_000        |
| `erc721`              |                  |            | 11.62 K              | 600_000        |
| `flipper`             |                  |            | 1.24 K               | 75_000         |
| `forward-calls`       |                  |            | 2.87 K               | 151_412        |
| `incrementer`         |                  |            | 1.14 K               |                |
| `mother`              |                  |            | 12.24 K              |                |
| `multisig`            |                  |            | 25.79 K              | 470_483        |
| `payment-channel`     |                  |            | 7.95 K               |                |
| `rand-extension`      |                  |            | 3.79 K               | 75_000         |
| `set-code-hash`       |                  |            | 1.49 K               |                |
| `subber`              |                  |            | 2.06 K               |                |
| `trait-erc20`         |                  |            | 8.69 K               | 225_000        |
| `trait-flipper`       |                  |            | 0.97 K               | 75_000         |
| `trait-incrementer`   |                  |            | 1.12 K               | 150_000        |
| `updated-incrementer` |                  |            | 9.78 K               |                |
| `upgradeable-flipper` |                  |            | 1.48 K               |                |



[Link to the run](https://gitlab.parity.io/parity/ink-waterfall/-/pipelines/209808) | Last update: Tue Aug 23 14:44:30 CEST 2022

**Review by ascjones — APPROVED** [review]
LGTM :heart:

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -201,6 +201,23 @@ check-wasm:
...
+    <<:                              *test-refs
+    script:
+    - cd linting/
+    # we are using a toolchain file in this directory
+    # add required components for CI
+    - rustup component add rustfmt clippy
+    - cargo check --verbose
+    - cargo fmt --verbose --all -- --check
+    - cargo clippy --verbose -- -D warnings;
+
```
Comment (alvicsam) [inline .gitlab-ci.yml:216]:
```suggestion
```
We tend not to divide the `script` section with empty lines in `.gitlab-ci.yml`. I can't say that it's a strict rule but would be nice to have the same format.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -201,6 +201,23 @@ check-wasm:
...
+    # Needed until https://github.com/mozilla/sccache/issues/1000 is fixed.
+    - unset RUSTC_WRAPPER
+
```
Comment (alvicsam) [inline .gitlab-ci.yml:219]:
```suggestion
```
The same here

File: linting/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,95 @@
+[package]
+name = "ink_linting"
+version = "4.0.0"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2021"
+publish = false
+
+license = "Apache-2.0"
+readme = "README.md"
+repository = "https://github.com/paritytech/cargo-contract"
```
Comment (HCastano) [inline linting/Cargo.toml:10]:
```suggestion
repository = "https://github.com/paritytech/ink"
```

File: linting/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+documentation = "https://docs.rs/ink_linting"
+homepage = "https://github.com/paritytech/cargo-contract"
```
Comment (HCastano) [inline linting/Cargo.toml:12]:
```suggestion
homepage = "https://github.com/paritytech/ink"
```

File: linting/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+path = "ui/pass/mapping-additional-logic-constructor.rs"
+
+[[example]]
+name = "pass_dont_use_fully_qualified_path"
+path = "ui/pass/mapping-dont-use-fully-qualified-path.rs"
+
+[package.metadata.rust-analyzer]
+rustc_private = true
+
+[workspace]
```
Comment (HCastano) [inline linting/Cargo.toml:83]: Do we need this?

**Review by HCastano — APPROVED** [review]

File: linting/Cargo.toml (hunk 5 above)
Comment (athei) [reply linting/Cargo.toml:83]: Yes. The driver does not build without that. All the examples in the dylint repo have this, too.

**Pull request merged**